test/functional/feature_taproot.py

```diff
@@ -0,0 +1,745 @@
+                # 13) OP_CHECKSIG with empty key that needs invalid signature
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_0, OP_CHECKSIG, OP_NOT]),
+                # 14) OP_CHECKSIGADD that needs invalid signature
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_0, pub1, OP_CHECKSIGADD, OP_NOT]),
+                # 15) OP_CHECKSIGADD with empty key that needs invalid signature
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_0, OP_0, OP_CHECKSIGADD, OP_NOT]),
+                # 16) OP_CHECKSIG with unknown pubkey type
+                CScript([OP_1, OP_CHECKSIG]),
+                # 17) OP_CHECKSIGADD with unknown pubkey type
+                CScript([OP_0, OP_1, OP_CHECKSIGADD]),
+                # 18) OP_CHECKSIGVERIFY with unknown pubkey type
+                CScript([OP_1, OP_CHECKSIGVERIFY, OP_1]),
+                # 19) script longer than 10000 bytes and over 201 non-push opcodes
+                CScript([OP_0, OP_0, OP_2DROP] * 10001 + [pub2, OP_CHECKSIG]),
+                # 20) OP_CHECKSIGVERIFY with empty key
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_0, OP_0, OP_CHECKSIGVERIFY, OP_1]),
+                # 21) Script that grows the stack to 1000 elements
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_1] + [OP_DUP] * 999 + [OP_DROP] * 999),
+                # 22) Script that grows the stack to 1001 elements
+                CScript([pub2, OP_CHECKSIGVERIFY, OP_1] + [OP_DUP] * 1000 + [OP_DROP] * 1000),
+                # 23) Script that expects an input stack of 1000 elements
+                CScript([OP_DROP] * 999 + [pub2, OP_CHECKSIG]),
+                # 24) Script that expects an input stack of 1001 elements
+                CScript([OP_DROP] * 1000 + [pub2, OP_CHECKSIG]),
+                # 25) Script that pushes a 520 byte element
+                CScript([random_bytes(520), OP_DROP, pub2, OP_CHECKSIG]),
+                # 26) Script that pushes a 521 byte element
+                CScript([random_bytes(521), OP_DROP, pub2, OP_CHECKSIG]),
+                # 27) Script that pushes a 521 byte element and OP_SUCCESSX
+                CScript([random_bytes(521), OP_DROP, pub2, OP_CHECKSIG, random_op_success()]),
+                # 28) Pushes random CScriptNum value, checks OP_CHECKSIGADD result
+                CScript([checksigadd_arg, pub2, OP_CHECKSIGADD, checksigadd_result, OP_EQUAL]),
+                # 30) CHECKSIGADD that succeeds with proper sig because numeric argument number is <=4 bytes
+                CScript([CScriptNum(oversize_number-1), pub2, OP_CHECKSIGADD]),
+                # 29) CHECKSIGADD that must fail because numeric argument number is >4 bytes
+                CScript([CScriptNum(oversize_number), pub2, OP_CHECKSIGADD]),
+            ]
+            # For the next test we must predict the exact witness size
+            witness_size = 141 + (hashtype != 0) + (0 if annex is None else len(annex) + 1)
```

Comment:
Wondering if you could un-magic this constant a bit. I was going to attempt to write another case for unknown pubkey using this template, but ran out of motivation to reverse engineer it.

Reply:
I still have to do this.

Reply:
Better now? I've extended the test further too, and moved it to its own section.